add media sharing document (#2683)
ably/docs: src/pages/docs/chat/rooms/media-sharing.mdx

---

> title: "Sharing media with Ably Chat"

Reviewer: I think this should be "Media sharing" to match with the nav and the format of other pages?

Author: done, thanks

meta_description: "How to share media like images, videos, or files with Ably Chat"
meta_keywords: "image, file, media, sharing, Ably Chat, chat SDK, realtime messaging, dependability, cost optimisation"
---

> A common requirement for chat applications is the ability to share media like images, videos, or files with other participants.

Reviewer: ```suggestion
A common requirement for chat applications is the ability to share media such as images, videos, or files.
```

'with others' is implicit in the sharing.

> Ably Chat currently does not support media sharing out of the box, but it is very straightforward to link media and metadata to messages, and this can be used to implement media sharing features.

Reviewer: This starts out a bit negative. Do we really need to state that we don't have a dedicated API for it?

Author: Changed it to a more positive tone, not mentioning we don't support it

> ## Architecture Overview

Reviewer: Sentence case for headings please.

Author: done, thanks

The general approach for implementing media sharing in Ably Chat is:

> 1. Upload media to your own storage service (your own servers, or a third-party service like AWS S3)

Reviewer: ```suggestion
1. Upload the media to your own storage service. This can be your own servers, or a third-party service like AWS S3.
```

> 2. Send a Chat Message with `metadata` containing information about the shared media

Reviewer: ```suggestion
2. Use the `metadata` field when [sending a message](/docs/chat/rooms/messages#send) to store the relevant information about the file.
```

Can we give examples here. I assume it's just the S3 URL for example?

3. On the receiving end, display the media when a message with the relevant metadata is received

## Implementation Guide

We'll build a simple chat room where users can share images with each other.

We assume that `room` is an Ably chat room that is attached. Please refer to our [getting started guide](/docs/chat/getting-started/javascript) for step-by-step instructions on how to set up Ably Chat and get a chat room.

We also assume that a function called `uploadImage() : Promise<string>` exists and handles the whole image upload flow: asks users to pick a picture, uploads it, and returns a unique identifier for this image.

> Let's say that the images will be served from `https://example.com/images/<identifier>`, where the identifier is a string of length between 10 and 15 and matches the `^[a-z0-9]{10,15}$` regex.

Reviewer: My view is that we don't need to treat this in a unique way just because there's some additional work required from the dev here. I think we should continue to make the headings action-based like for any other feature, and this would just become something like 'Create an upload function' where we state that this is something that they need to write and handle themselves for whichever service they're using and provide an example as you have done.

Author: I've changed the style, thanks

> ### Sending messages with images

Reviewer: This would become H2 depending on my above comment. But stick to action-orientated headings: `Send a message with an X` 

I say `X` because I think we should be consistent throughout this page in the terminology we use. Is it 'media', 'an attachment', 'a file'.

Author: changed them

> When the user initiates the upload flow (for example by clicking a button called "attach"), we use the `uploadImage()` flow and save the resulting ID, like this:

Reviewer: This doesn't read the best. Can we not just state "When a user uploads/attaches a file..." 

Please avoid using the first person too (no "we do this").

> ```javascript

Reviewer: `<Code>` tags needed here, yep.

let imagesToAttach = [];
async function onImageAttach() {
  const imageId = await uploadImage();
  imagesToAttach.push(imageId);
}
```

> In the UI, the `imagesToAttach` array is displayed so the users know what images will be attached to their message. The UI can allow users to remove or sort the images.

Reviewer: ```suggestion
In your UI, the `imagesToAttach` array should be displayed so that users know which images will be attached to their message. It can allow users to remove or sort the images.
```

Author: applied your version, thanks!

    return message.metadata.images.filter(imageId => imageIdRegex.test(imageId));
  }
  return [];
}
```

And now for each message we can display images if it has any.
We provide here a simplistic example for displaying text and images by constructing the DOM for them manually. The same high-level logic applies for modern JavaScript front-end frameworks, Swift, and Kotlin.

```javascript
function createMessageDOM(message) {
  const container = document.createElement("div");
  const text = document.createElement("div");
  text.innerText = message.text;
  container.appendChild(text);
  const validImages = getValidImages(message);
  if (validImages.length > 0) {
    const imagesContainer = document.createElement("div");
    for (let imageId of validImages) {
      const img = document.createElement("img");
      img.src = "https://example.com/images/"+imageId;
      imagesContainer.appendChild(img);
    }
    container.appendChild(imagesContainer);
  }
  return container;
}
```

### Why image ID and not URL?

Image IDs are simpler to validate than full URLs. You can also use URLs, but you must correctly validate them on the receiving end.

### Metadata for each image

> In the example above we used a simple string array to store image IDs. We can change this to be an object where each image holds extra information that can be displayed to the user, such as a title or description, or the expected size in pixels so that the UI can allocate a placeholder loading box of the correct proportions.

Reviewer: Should this be titled differently and provide an example?

Author: I've removed this and instead made it the default.

### Add an image to an existing message

You can add an image to an existing message by editing its metadata.

```javascript
let imageId = 'abcd123abcd'; // assume this is the image we want to add
let message = ...; // assume this is the message we want to edit

const newMetadata = structuredClone(message.metadata);
if (!newMetadata.images) {
  newMetadata.images = [];
}
newMetadata.images.push(imageId);

room.messages.update(message.serial, message.copy({metadata: newMetadata}))
```

### Remove an image from an existing message

Similarly to how we add a new image, we can remove existing images from messages by updating the message:

```javascript
let imageId = 'abcd123abcd'; // assume this is the image we want to remove
let message = ...; // assume this is the message we want to edit

if (!message.metadata.images || message.metadata.images.length === 0) {
  //do nothing if there are no images
  return; 
}
const newMetadata = structuredClone(message.metadata);
newMetadata.images = newMetadata.images.filter(id => imageId !== id);

room.messages.update(message.serial, message.copy({metadata: newMetadata}))
```

## Privacy and security considerations

> The message metadata is not validated by the server. Always treat it as untrusted user input. The same applies to the text and headers of a chat message.

Reviewer: This could be a note when we first mention using the metadata field.

Author: Moved

### Validate metadata on the receiving end

> You must always validate the metadata before displaying any media to the user. Treat the metadata as untrusted user input. If you are using IDs, validate they are of the correct format. If you are using URLs, validate the schema, domain, path and query parameters are what you expect. Do not display images or other media directly from the metadata of a message.

Reviewer: This also seems like something to mention within the "Display images/attachments" section.

> ### Use authentication server-side to control access

Reviewer: Should this be higher up the page? Seems important to understand the access controls before setting it up.
